ambry-store/src/main/java/com/github/ambry/store/IndexSegment.java

```diff
@@ -1010,11 +1009,37 @@ private void readFromFile(File fileToRead, Journal journal) throws StoreExceptio
     }
   }
 
+  /**
+   * Check data integrity of index file (represented by byte buffer). This methods computes crc of byte buffer and
+   * compares it with crc value at the end of file.
+   * @throws StoreException
+   */
+  private void checkDataIntegrity() throws StoreException {
+    byte[] byteArray;
+    if (serEntries.hasArray()) {
+      // this is for heap buffer only
+      byteArray = serEntries.array();
+    } else {
+      // this is for non-heap (direct) buffer
+      serEntries.position(0);
+      byteArray = new byte[serEntries.capacity()];
+      serEntries.get(byteArray, 0, serEntries.capacity());
+    }
+    Crc32 crc = new Crc32();
+    crc.update(byteArray, 0, byteArray.length - CRC_FIELD_LENGTH);
```

Comment:
actually you don't have to copy bytebuffer to an byte array, Crc32 has a method to update CRC with ByteBuffer directly and it doesn't matter if it's heap bytebuffer or direct bytebuffer.

Reply:
True, there is a method to update CRC with ByteBuffer. However, we should exclude the last 8 bytes in buffer to calculate CRC, right?  How can we do that with that method?

Reply:
You can do this by setting the limit on the ByteBuffer before calling the CRC method.

Reply:
Thank you @justinlin-linkedin and @cgtz , the code should be more concise now.